### PR TITLE
Add status column to featured_projects table

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds the `status` column to the `featured_projects` table in anticipation of changes to the public project gallery.
This is part 1 of 3 steps described [here](https://codedotorg.atlassian.net/browse/LABS-497).

## Links
[jira](https://codedotorg.atlassian.net/browse/LABS-497)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/featured_project.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #  featured_at    :datetime
 #  unfeatured_at  :datetime
 #  topic          :string(255)
+#  status         :string(255)      default("saved")
```
Comment (bencodeorg) [inline dashboard/app/models/featured_project.rb:10]: Will we ever want to do something like "return all projects with status = saved"? If so, we probably want an index on it, like we have on topic

**Review by bencodeorg — COMMENTED** [review]
The little diagram in the Jira ticket is very helpful! One clarification -- what is the distinction between the "archived" and "featured" statuses and the current featured / unfeatured implementation?

File: dashboard/app/models/featured_project.rb (hunk 1 above)
Comment (fisher-alice) [reply dashboard/app/models/featured_project.rb:10]: Thanks. Do I just add the following at line 16?
```
index_featured_projects_on_status           (status)
```

**Comment by fisher-alice** [comment]
> The little diagram in the Jira ticket is very helpful! One clarification -- what is the distinction between the "archived" and "featured" statuses and the current featured / unfeatured implementation?

"featured" status will denote projects that are currently featured at '/projects/public'.
"archived" status will denote projects that were featured at some point in the past and no longer featured.
"saved" status is the default. Instead of marking a project as "featured" immediately, an admin user will be able to mark a project as "saved". Then at "projects/featured" the admin user can update a project's status to "featured".
There will be 3 tables displayed at 'projects/featured'.

So adding the index on "status" is key. Thanks for that!

And @samantha-code created the diagram after we discussed the updates. Thanks Sam!

File: dashboard/app/models/featured_project.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/app/models/featured_project.rb:10]: No, you need to update your migration to add an index, something like this:

https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/db/migrate/20210126021131_add_topic_to_featured_projects.rb

I believe you can rerun it by doing something like `bundle exec rake db migrate:down` then `bundle exec rails db migrate` again.

I'm not sure about the character set step though in this migration, that's a new one to me...looks like newer migrations aren't doing that, so I think you can ignore.

**Comment by bencodeorg** [comment]
Maybe to clarify my question -- are the featured/archived states any different than what is currently derived from the featured_at / unfeatured_ at timestamps?

**Comment by fisher-alice** [comment]
> Maybe to clarify my question -- are the featured/archived states any different than what is currently derived from the featured_at / unfeatured_ at timestamps?

Gotcha. The 'featured' state will continue to be derived as it currently is [here](https://github.com/code-dot-org/code-dot-org/blob/0939e83d482ef04c26714bb75cace8e3534c2011/dashboard/app/models/featured_project.rb#L24-L26).

In a follow-up PR, I will add a `save` method in `featured_projects_controller` and a new featured project's `featured_at` and `unfeatured_at` properties will be assigned `NULL`.

| status | featured_at | unfeatured_at |
| -- | -- | -- |
| 'saved' | NULL | NULL |
| 'featured' | most recently featured at time stamp | NULL |
| 'archived' | most recently featured at time stamp | most recently unfeatured at time stamp |

Does this answer your question? thanks!

**Comment by bencodeorg** [comment]
It seems like your desired statuses can be achieved without a new column? Like you describe in the table in your comment, each state can be derived from the combination of featured_at / unfeatured_at. The new "saved" state is just a row with null for both timestamps?

**Comment by fisher-alice** [comment]
> It seems like your desired statuses can be achieved without a new column? Like you describe in the table in your comment, each state can be derived from the combination of featured_at / unfeatured_at. The new "saved" state is just a row with null for both timestamps?

That's a great point! I'm glad you asked that question which led to the table above 😄 . 

You brought up indexing by status. I believe we'll still be able to filter featured projects into 3 groups depending on their `featured_at` and `unfeatured_at` properties. Could you explain indexing and why we need it for 'topic'?

**Comment by fisher-alice** [comment]
Turns out adding `status` column is not needed when we add a third category of featured projects 'saved'.

**Pull request closed without merging**